cip/CIP2016-06-22-nested-subqueries.adoc

```diff
@@ -0,0 +1,347 @@
+  RETURN tweet, favourite.time AS time, user.country AS country
+}
+WHERE country = 'se'
+RETURN DISTINCT tweet
+ORDER BY time DESC
+LIMIT 10
+----
+
+Uncorrelated nested subquery:
+[source, cypher]
+----
+MATCH (f:Farm {id: $farmId})
+MATCH {
+  MATCH (u:User {id: $userId})-[:LIKES]->(b:Brand),
+        (b)-[:PRODUCES]->(p:Lawnmower)
+  RETURN b.name AS name, p.code AS code
+  UNION
+  MATCH (u:User {id: $userId})-[:LIKES]->(b:Brand),
+        (b)-[:PRODUCES]->(v:Vehicle),
+        (v)<-[:IS_A]-(:Category {name: 'Tractor'})
+  RETURN b.name AS name, p.code AS code
+}
+RETURN f, name, code
+----
+
+Correlated nested subquery:
+[source, cypher]
+----
+MATCH (f:Farm {id: $farmId})-[:IS_IN]->(country:Country)
+MATCH {
+  MATCH (u:User {id: $userId})-[:LIKES]->(b:Brand),
+        (b)-[:PRODUCES]->(p:Lawnmower)
+  RETURN b.name AS name, p.code AS code
+  UNION
+  MATCH (u:User {id: $userId})-[:LIKES]->(b:Brand),
+        (b)-[:PRODUCES]->(v:Vehicle),
+        (v)<-[:IS_A]-(:Category {name: 'Tractor'})
+  WHERE v.leftHandDrive = country.leftHandDrive
+  RETURN b.name AS name, p.code AS code
+}
+RETURN f, name, code
+----
+
+Filtered and correlated nested subquery:
+[source, cypher]
+----
+MATCH (f:Farm)-[:IS_IN]->(country:Country)
+WHERE country.name IN $countryNames
+MATCH {
+  MATCH (u:User {id: $userId})-[:LIKES]->(b:Brand),
+        (b)-[:PRODUCES]->(p:Lawnmower)
+  RETURN b AS brand, p.code AS code
+  UNION
+  MATCH (u:User {id: $userId})-[:LIKES]->(b:Brand),
+        (b)-[:PRODUCES]->(v:Vehicle),
+        (v)<-[:IS_A]-(:Category {name: 'Tractor'})
+  WHERE v.leftHandDrive = country.leftHandDrive
+  RETURN b AS brand, p.code AS code
+}
+WHERE f.type = 'organic'
+  AND b.certified
+RETURN f, brand.name AS name, code
+----
+
+Doubly-nested subquery:
+[source, cypher]
+----
+MATCH (f:Farm {id: $farmId})
+MATCH {
+  MATCH (c:Customer)-[:BUYS_FOOD_AT]->(f)
+  MATCH {
+     MATCH (c)-[:RETWEETS]->(t:Tweet)<-[:TWEETED_BY]-(f)
+     RETURN c, count(*) AS count
+     UNION
+     MATCH (c)-[:LIKES]->(p:Posting)<-[:POSTED_BY]-(f)
+     RETURN c, count(*) AS count
+  }
+  RETURN c, 'customer' AS type, sum(count) AS endorsement
+  UNION
+  MATCH (s:Shop)-[:BUYS_FOOD_AT]->(f)
+  MATCH (s)-[:PLACES]->(a:Advertisement)-[:ABOUT]->(f)
+  RETURN s, 'shop' AS type, count(a) * 100 AS endorsement
+}
+RETURN f.name AS name, type, sum(endorsement) AS endorsement
+----
+
+**2. Write-only/read-write subqueries**
+
+We illustrate these by means of an 'old' version of the query, in which `FOREACH` is used, followed by the 'new' version, using `DO`.
+
+Using a single subquery - old version using `FOREACH`:
+[source, cypher]
+----
+MATCH (r:Root)
+FOREACH(x IN range(1, 10) |
+  MERGE (c:Child {id: x})
+  MERGE (r)-[:PARENT]->(c)
+)
+----
+
+Using a single subquery - new version using `DO`:
+[source, cypher]
+----
+MATCH (r:Root)
+UNWIND range(1, 10) AS x
+DO {
+  MERGE (c:Child {id: x})
+  MERGE (r)-[:PARENT]->(c)
+}
+----
```

Comment:
Might be nice to explain the difference between the preceding example and what it would be without the `DO` and just having the `MERGE` calls following `UNWIND`. I assume it's simply keeping the same cardinality (i.e. the `DO` itself has a cardinality of 1) and not introducing any new identifiers, but it might be helpful to be explicit about that and why it's helpful.